Add Smart Payment Buttons mode as alternative to directly embedded image links

dechov opened the pull request:
<img width="510" alt="screen shot 2018-06-01 at 4 47 19 pm" src="https://user-images.githubusercontent.com/1867547/40864341-6d0a393e-65c1-11e8-9337-8b7c701ad1f0.png">

(Toggled off by default for upgrading merchants, and on for new installs.)

Front-end view | Relevant settings | Fixes
-- | -- | --
Cart: <img width="434" alt="screen shot 2018-06-01 at 9 57 57 am" src="https://user-images.githubusercontent.com/1867547/40844561-49774f46-6582-11e8-96f1-1b452650311a.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/99305a6524c65dc77f995febfd58dcfc18f4de7c | <img width="671" alt="screen shot 2018-06-01 at 9 56 55 am" src="https://user-images.githubusercontent.com/1867547/40844579-57a45f32-6582-11e8-8bed-b77b2497a8d9.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/ea578d97abe0a2a41b3d7c910d932cdaf8e225c9 https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/2ca813a3256ccce9ec79a824b5b493992925750f https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/2318eb17eff01078a152a5ad734c7d821b6f975f https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/a9f4dec8b207e11548f377931ede72af0294190e
Single Product: <img width="529" alt="screen shot 2018-06-01 at 9 10 31 am" src="https://user-images.githubusercontent.com/1867547/40844278-8170d832-6581-11e8-9e1b-f8cbd2386475.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/d3727d6282b90a61c3aa578305a9406bbf9d619d | <img width="669" alt="screen shot 2018-06-01 at 9 55 10 am" src="https://user-images.githubusercontent.com/1867547/40844435-e8f644ec-6581-11e8-998a-0e0f408a8683.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/428ab6eb2a6ac34a6f624c91e1ae30b6622ff705 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/407
Checkout: <img width="346" alt="screen shot 2018-06-01 at 10 06 08 am" src="https://user-images.githubusercontent.com/1867547/40845037-7c82abfa-6583-11e8-9f4e-a781cb4e9903.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/baf4f69cd29bebac51d95bc5445559d88f0f1ddd | <img width="664" alt="screen shot 2018-06-01 at 10 07 09 am" src="https://user-images.githubusercontent.com/1867547/40845063-90670300-6583-11e8-8aa6-2f37cff5ff0a.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/b1137aa36469ecceb15c8df0ce9bc698ddb37d87 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/403
Mini-cart: <img width="231" alt="screen shot 2018-06-01 at 10 09 32 am" src="https://user-images.githubusercontent.com/1867547/40845183-ea9db706-6583-11e8-868d-c0f7f17292f7.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/7ac071be3a4096a37f77368e5bda44b8e9dda90a | <img width="666" alt="screen shot 2018-06-01 at 10 10 04 am" src="https://user-images.githubusercontent.com/1867547/40845196-fc2e2492-6583-11e8-873b-972d9a65de5a.png"> https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/019648e37a8169311b57bdd8fedd2068a3d971b1 | https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/328

dechov commented:
Note: in the existing JS code for the single product page, a `generate_cart` call is made before redirecting to PayPal. With the new integration, this could _probably_ just as well be done during the `start_checkout` call, but I've left it as a separate call preceding it so as to reduce the potential for unforeseen side effects. Open to changing it.

Another thing I should note: although little or no code is actually removed in this PR, a future update will remove the legacy mode, and though it's not yet precisely clear what we can remove at that point, there will likely be a large amount to clean up.

jeffstieler commented on includes/class-wc-gateway-ppec-cart-handler.php:
```diff
@@ -223,11 +299,68 @@ public function enqueue_scripts() {
+					);
+				}
+
+				$data = array_merge( $data, $this->convert_to_render_settings( $page_data ) );
```
Since these 4 settings are repeated a lot, and are the only ones inspected/modified in `convert_to_render_settings()` - what do you think about combining the retrieval of them with an optional prefix from `$settings` and the "conversion" into a single method? Something like `get_button_settings( $settings, 'single_product' )`?

dechov replied:
Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/86932eefb55ab497b11d51c4714b36a3aa58327d

jeffstieler commented on includes/class-wc-gateway-ppec-with-spb.php:
```diff
@@ -0,0 +1,59 @@
+	 * @return array
+	 */
+	public function process_payment( $order_id ) {
+		if ( isset( $_POST['payerID'] ) ) {
```
We should probably check for `paymentToken` here too.

jeffstieler commented on includes/class-wc-gateway-ppec-with-spb.php:
```diff
@@ -0,0 +1,59 @@
+	 *
+	 * @since 1.6.0
+	 */
+	public function payment_scripts() {
```
Can you document this behavior with a comment?

jeffstieler commented on includes/class-wc-gateway-ppec-with-spb.php:
```diff
@@ -0,0 +1,59 @@
+	 *
+	 * @return array
+	 */
+	public function process_payment( $order_id ) {
```
Can you document this behavior with a comment?

jeffstieler commented on includes/settings/settings-ppec.php:
```diff
@@ -107,13 +107,115 @@
+			} else if ( ! button_size.val() ) {
+				button_size.val( 'large' ).change();
+			}
+		} ).change();
```
All of this conditional form input logic needs documenting.

jeffstieler commented on includes/settings/settings-ppec.php:
```diff
@@ -400,4 +457,168 @@
+		'description' => __( 'Customize the appearance of Express Checkout on your site.', 'woocommerce-gateway-paypal-express-checkout' ),
+	),
+	'use_spb' => array(
+		'title'       => __( 'Smart Payment Buttons', 'woocommerce-gateway-paypal-express-checkout' ),
```
Duplicate key.

dechov replied:
Thanks – not sure when or how the latter one got changed from `label` to `title`.

jeffstieler commented on includes/settings/settings-ppec.php:
```diff
@@ -400,4 +457,168 @@
+foreach( $per_context_settings as $key => $value ) {
+	$value['class'] .= ' woocommerce_ppec_paypal_mark';
+	$settings[ 'mark_' . $key ] = $value;
+}
```
A comment breaking up each "context" would make this much more readable.

dechov commented on assets/js/wc-gateway-ppec-smart-payment-buttons.js:
```diff
@@ -0,0 +1,99 @@
+;( function ( $, window, document ) {
+	'use strict';
+
+	var getFundingMethods = function( methods ) {
```
True, but `_` isn't available here – would you suggest loading it?

dechov replied:
Did the latter in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/9121cc64dc3f2394bb4221b9cb77f1c756f68e5b

dechov commented on includes/class-wc-gateway-ppec-cart-handler.php:
```diff
@@ -118,6 +119,24 @@ public function wc_ajax_update_shipping_costs() {
+			add_filter( 'woocommerce_cart_needs_shipping', '__return_false' );
+		}
+
+		wc_gateway_ppec()->checkout->start_checkout_from_cart();
```
Good question – will test and make sure errors are handled and reported.

dechov commented:
Added comments and addressed other feedback. Let me know if there's any more code that's difficult to read or understand.

Also enabled SPB mode by default for new installs, and added an admin notice when enabled but not in SPB mode:

<img width="1022" alt="screen shot 2018-06-18 at 10 20 20 pm" src="https://user-images.githubusercontent.com/1867547/41573329-2fcccb22-734a-11e8-93a9-a137a73b85bd.png">

Still need to add descriptions but opening back up to code review in the meantime.

jeffstieler commented on includes/settings/settings-ppec.php:
```diff
@@ -499,7 +499,7 @@ function showHideDefaultButtonSettings() {
 		'title'       => __( 'Smart Payment Buttons', 'woocommerce-gateway-paypal-express-checkout' ),
 		'type'        => 'checkbox',
-		'default'     => 'no',
+		'default'     => $this->get_option( 'button_size' ) ? 'no' : 'yes', // Hinge on whether settings have been initialized.
```
Nit: being more specific in this comment would be helpful. E.g. `button_size` being set indicates an existing PPEC install.

dechov replied:
Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/383872cdbdb793ec8d7edca591aeb5a851385fba

dechov commented on includes/class-wc-gateway-ppec-plugin.php:
```diff
@@ -222,6 +226,29 @@ public function show_bootstrap_warning() {
+			<script>
+			( function( $ ) {
+				$( '.ppec-dismiss-spb-notice' ).on( 'click', '.notice-dismiss', function() {
+					jQuery.post( "<?php echo admin_url( 'admin-ajax.php' ); ?>", {
```
Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/5113019867a9e4b744aeb1645e5e4a5d244bf9c8

dechov commented on includes/class-wc-gateway-ppec-plugin.php:
```diff
@@ -227,9 +225,22 @@ public function show_bootstrap_warning() {
+		// Should only show when PPEC is enabled but not in SPB mode.
+		if ( 'yes' !== $this->settings->enabled || 'yes' === $this->settings->use_spb ) {
+			return;
+		}
```
Done in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/9a8e6e569846e24ba9baeb1f8b6194f6cf8426ba

jeffstieler commented on includes/settings/settings-ppec.php:
```diff
@@ -521,6 +523,7 @@ function showHideDefaultButtonSettings() {
 		'class'       => 'wc-enhanced-select woocommerce_ppec_paypal_spb',
 		'default'     => 'rect',
 		'desc_tip'    => true,
+		'description' => __( 'The pill-shaped button\'s unique and powerful shape signifies PayPal in people’s minds. Use the rectangular button as an alternative when pill-shaped buttons might pose design challenges.', 'woocommerce-gateway-paypal-express-checkout' ),
```
Intentional usage of both `'` and `’` ?

dechov replied:
Nice catch – changed in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/409/commits/088b59cff622190aa7a2f413b5e55b579c2348a8.

othqsa95 commented:
Hello everyone,

I added the PAYPAL button in the MINI-CART and I adjusted it well with the CSS. I note however, that when I add a product to the basket, the slide cart opens and the button is displayed correctly.

But when I leave the slide cart and open it a second time without adding new products, the Paypal button is not displayed.

Anyone have an idea to solve this little problem?
